src/pyrokinetics/diagnostics/__init__.py

```diff
@@ -248,3 +251,199 @@ def _invfft(f, x, y, kx, ky):
+
+        grad_alpha = np.sqrt(g_aa)
+
+        g_tt = metric.field_aligned_covariant_metric("theta", "theta")
+        g_rt = metric.field_aligned_covariant_metric("r", "theta")
+        g_rr = metric.field_aligned_contravariant_metric("r", "r")
+
+        g_rt_contr = metric.toroidal_contravariant_metric("r", "theta")
+        g_rr_contr = metric.toroidal_contravariant_metric("r", "r")
+
+        # Jacobian and derivatives
+        jacob = metric.Jacobian
+
+        # R and derivatives
+        R = metric.R
+        dR_dr = metric.dRdr
+
+        # Z and derivatives
+        Z = metric.Z
+        dZ_dr = metric.dZdr
+
+        # B_mag and derivatives
+        B_mag = metric.B_magnitude
+        dB_dr = metric.dB_magnitude_dr
+        dB_dtheta = metric.dB_magnitude_dtheta
+
+        # Drifts
+        gbdrift = -2 * (dB_dr - g_rt / g_tt * dB_dtheta) / B_mag
+        press = -2 * dpdrho * (jacob / np.sqrt(g_tt)) / B_mag / dpsidrho
+        gbdrift0 = 2 * metric.dqdr * g_at / g_tt * dB_dtheta / B_mag
+        cvdrift = gbdrift + press
+
+        # GDS values
+        gds2 = (grad_alpha * dpsidrho) ** 2
+        gds21 = -(dpsidrho**2) * g_ra_contr * metric.dqdr
+        gds22 = (dpsidrho * metric.dqdr) ** 2 * g_rr
+
+        # Parallel gradient
+        gradpar = 1 / np.sqrt(g_tt)
+
+        # B field
+        bmag = B_mag
+        bpol = dpsidrho * np.sqrt(g_rr) / R
+
+        # Grad rho
+        grho = np.sqrt(g_rr)
+
+        # Eikonal
+        S = -metric.alpha
+        dS_dr = (
+            -dpsidrho
+            / np.sqrt(g_rr_contr)
+            * (metric.dalpha_dr * g_rr_contr + metric.dalpha_dtheta * g_rt_contr)
+        )
+
+        return {
+            "dpdrho": dpdrho,
+            "theta": theta,
+            "bmag": bmag,
+            "bpol": bpol,
+            "gradpar": gradpar,
+            "cvdrift": cvdrift,
+            "gds2": gds2,
+            "gbdrift": gbdrift,
+            "gbdrift0": gbdrift0,
+            "grho": grho,
+            "gds21": gds21,
+            "gds22": gds22,
+            "rplot": R,
+            "zplot": Z,
+            "rprime": dR_dr,
+            "zprime": dZ_dr,
+            "aplot": S,
+            "aprime": dS_dr,
+        }
+
+    def ideal_ballooning_solver(self):
+        r"""
+        Adapted from ideal-ballooning-solver
+        https://github.com/rahulgaur104/ideal-ballooning-solver/blob/master/ball_scan.py
+
+        """
+
+        geometry_terms = self.gs2_geometry_terms()
+
+        dpdrho = geometry_terms["dpdrho"]
+        theta = geometry_terms["theta"]
+        bmag = geometry_terms["bmag"]
+        gradpar = geometry_terms["gradpar"]
+        cvdrift = geometry_terms["cvdrift"]
+        gds2 = geometry_terms["gds2"]
+
+        gamma, X_arr, dX_arr, g_arr, c_arr, f_arr = gamma_ball_full(
+            dpdrho, theta, bmag, gradpar, cvdrift, gds2
+        )
+
+        return gamma
+
+
+def gamma_ball_full(
+    dPdrho, theta_PEST, B, gradpar, cvdrift, gds2, vguess=None, sigma0=0.42
```

Comment:
If one takes gds21, gds22 and cvdrift0 then you should be able to include the theta0 dependence here trivially I think with `gds2 --> gds2 + 2 * theta0 * gds21 + theta0**2 * gds22` and `cvdrift --> cvdrift + theta0 * cvdrift0`

Reply:
We can add that to the function here and give `theta0` as an argument. Seems the best way to handle this.

Reply:
Do any other terms change with `theta0`? Maybe put this into `get_gs2_geometry` is more appropriate?

Reply:
Np. For ballooning and local gyrokinetic solvers, it's just gds2 and cvdrift. The function here assumes that you are adding the theta0 terms before passing gds2 and cvdrift to the solver.

Reply:
Okay I'll put it in `get_gs2_geometry` then!